packages/test/test-end-to-end-tests/src/test/container.spec.ts

```diff
@@ -857,3 +858,95 @@ describeCompat("Driver", "NoCompat", (getTestObjectProvider) => {
 		assert.equal(storage.policies?.maximumCacheDurationMs, fiveDaysMs);
 	});
 });
+
+describeCompat("Container connections", "NoCompat", (getTestObjectProvider) => {
+	let provider: ITestObjectProvider;
+	let clock;
+	before(() => {
+		clock = useFakeTimers();
+	});
+	beforeEach("", async function () {
+		provider = getTestObjectProvider();
+		if (provider.driver.type !== "local") {
+			this.skip();
+		}
+	});
+	afterEach(() => {
+		clock.reset();
+	});
+	after(() => {
+		clock.restore();
+	});
+	it("container disconnect() stops the connection re-attempt loop", async () => {
+		let emulateThrowErrorOnConnection = false;
+		const retryAfter = 3;
+		let reconnectionAttemptCount = 0;
+		(provider as any)._documentServiceFactory = wrapObjectAndOverride<IDocumentServiceFactory>(
+			provider.documentServiceFactory,
+			{
+				createDocumentService: {
+					connectToDeltaStream: (_ds) => async (client) => {
+						// We let the container get created first before starting emulate throwing of errors.
+						if (emulateThrowErrorOnConnection) {
+							reconnectionAttemptCount++;
+							throw new RetryableError("Test message", "ThrottlingError", {
+								retryAfterSeconds: retryAfter,
+								driverVersion: "1",
+							});
+						} else {
+							return _ds.connectToDeltaStream(client);
+						}
+					},
+				},
+			},
+		);
+		// Create container
+		const container = await provider.makeTestContainer();
+		await waitForContainerConnection(container);
+		emulateThrowErrorOnConnection = true;
+
+		// This flag will ensure that the container warnings were observed when throttling error was thrown
+		let didReceiveContainerWarning = false;
+
+		// Host apps can chose to listen to container warning events and disconnect the container if they observe throttling errors
+		container.once("warning", async (warning) => {
```

Comment:
event callbacks can't be async. I'd guess that's the reason build fails.
You can still go with the same flow, just make it synchronous and use Promise.resolve().then() kind of flow (or queueMicrotask(), which seems more elegant way to do the same).

But I'd restructure it. As written, test only validates that we went as far as line 913 (setting didReceiveContainerWarning). If it took too long to get there, then we may exit test long before this callback continues past line 925.